consensus/src/block_preparer.rs

```diff
@@ -53,7 +55,14 @@ impl BlockPreparer {
         // Transaction filtering, deduplication and shuffling are CPU intensive tasks, so we run them in a blocking task.
         tokio::task::spawn_blocking(move || {
             let filtered_txns = txn_filter.filter(block_id, block_timestamp_usecs, txns);
-            let deduped_txns = txn_deduper.dedup(filtered_txns);
+            let mut deduped_txns = txn_deduper.dedup(filtered_txns);
+            let num_txns_before = deduped_txns.len();
+            deduped_txns.retain(|txn| {
+                Duration::from_micros(block_timestamp_usecs).as_secs()
+                    < txn.expiration_timestamp_secs()
+            });
+            NUM_EXPIRED_TXNS_IN_BLOCK_PREPARER
+                .observe((num_txns_before - deduped_txns.len()) as f64);
```

Comment:
maybe we should put this inside TransactionFilter? (makes deduper also cheaper)

i.e. do it before this short circuit line here - https://github.com/aptos-labs/aptos-core/blob/main/consensus/src/transaction_filter/mod.rs#L23


@sitalkedia for thoughts, and also for whether this should be a local config

Reply:
+1, better to put this insight the existing filter..

Reply:
I am currently counting the expired transactions in block preparer and proof queue and added some metrics for them.
I can alternately add the filtering in TransactionFilter, but I won't be able to add metrics. Is that fine?

Reply:
why wouldn't you be able to have a counter there?

Reply:
TransactionFilter consists of a vector of rules, such as deny txns from a particular sender, deny txns to a particular entry function, etc. It executes these rules one by one on each of the txns.

I can add a rule to deny expired txns in TransactionFilter. As the transaction filter is designed to be generic, it may not look aesthetically nice to have a counter to count the number of times this particular rule returned true. If we are okay with not having the counter, I can do the filter in TransactionFilter.

Reply:
I meant applying on this line, before any rule is called

https://github.com/aptos-labs/aptos-core/blob/main/consensus/src/transaction_filter/mod.rs#L23

also seems like filter + collect below should be changed to retain as well

Reply:
Ah. Gotcha. Moved the filter inside TransactionFilter. For efficiency, replaced filter + collect with retain and updated the test cases.